pkgs/stdenv/freebsd/make-bootstrap-tools.nix

```diff
@@ -0,0 +1,102 @@
+            cat $f >>"$base/nix-support/$f"
+          done
+        done
+        cd $base
+
+        ${fixups}
+
+        rm .nix-socket
+        ${packCmd}
+      '');
+    nar-all = pack-all "nix-store --dump . | xz -9 -e -T $NIX_BUILD_CORES >$out";
+    tar-all = pack-all "XZ_OPT=\"-9 -e -T $NIX_BUILD_CORES\" tar cJf $out --hard-dereference --sort=name --numeric-owner --owner=0 --group=0 --mtime=@1 .";
+    coreutils-big = pkgs.coreutils.override { singleBinary = false; };
+    mkdir = runCommand "mkdir" { coreutils = coreutils-big; } ''
+      mkdir -p $out/bin
+      cp $coreutils/bin/mkdir $out/bin
+    '';
+  in rec {
+  unpack = nar-all "unpack.nar.xz" (with pkgs; [bash mkdir xz gnutar]) ''
+    rm -rf include lib/*.a lib/i18n lib/bash share
+  '';
+  bootstrap-tools = tar-all "bootstrap-tools.tar.xz" (
+    with pkgs;
+    [
+      (runCommand "bsdcp" { } "mkdir -p $out/bin; cp ${freebsd.cp}/bin/cp $out/bin/bsdcp")
+      coreutils
+      gnutar
+      findutils
+      gnumake
+      gnused
+      patchelf
+      gnugrep
+      gawk
+      diffutils
+      patch
+      bash
+      xz
+      xz.dev
+      gzip
+      bzip2
+      bzip2.dev
+      curl
+      expand-response-params
+      binutils-unwrapped
+      freebsd.libc
+      llvmPackages.libcxx
+      llvmPackages.libcxx.dev
+      llvmPackages.compiler-rt
+      llvmPackages.compiler-rt.dev
+      llvmPackages.clang-unwrapped
+      (freebsd.locales.override { locales = [ "C.UTF-8" ]; })
+    ]
+    # INSTRUCTIONS FOR GENERATING THE SPURIOUS LIST
+    # - empty this list
+    # - rebuild bootstrap files and update their urls and hashes
+    # - turn on atime on your FreeBSD nix store filesystem
+    # - run nix-collect-garbage on FreeBSD to make it so we rebuild FODs
+    # - build the nixpkgs __bootstrapArchive attribute on FreeBSD
+    # - reboot your FreeBSD system. Otherwise the atimes will simply be wrong because of kernel caching
+    # - run a full build of stdenv on FreeBSD. with -j3, this takes 1h40 on my 20 cpu VM (AMD host)
+    # - use the following to generate a list with access times and filenames
+    #   find /nix/store/###-bootstrap-archive -type f | xargs stat | grep -E 'Access: 2|File:' | paste -d ' ' - - | awk '{ print $4 " " $5 " " $6 " " $2 }' | sort -n > atimes
+    # - manually identify the point where files have no longer been accessed after the patching phase
+    # - use your favorite text editor to snip out the time column, the /nix/store/###-bootstrap-archive/ prefix, and the files that have not been used during bootstrap
+    # - turn off atime if it was off before since it will degrade performance
+    # - manually remove bin/strings from the list, since it will be used only during bootstrap
+    # - manually remove all files under include and lib/clang/*/include from the list in order to improve forward compatibility (and since they are very small)
+    # - plop it here
+  ) "xargs rm -f <${./bootstrap-tools-spurious.txt}";
```

Comment:
Would be nice if this step was not as manual as this one is, but it's just an optimization after all. Potentially this could be also made nicer using dtrace on the process group that does all the building rather than relying on atime.

Reply:
this is more or less what i did on darwin then added back the clang header files.

```sh
#!/usr/bin/env nix-shell
#! nix-shell -i bash -p nix sqlite --pure
set -eou pipefail
# add some whitespace to trigger a rebuild
echo >> pkgs/stdenv/freebsd/unpack-bootstrap-files.sh

# this at least will trigger a rebuild using new tools on darwin
nix-build -A freshBootstrapTools.test-pkgs.stdenv

# create a reference file with a timestamp 1 second greater than the
# registration time for the derivation
name="bootstrap-archive"
touch  reference --date=@$((
  1 +
  $(sqlite3 -readonly /nix/var/nix/db/db.sqlite  <<SQL
select
  registrationTime
from
  ValidPaths
where
  path glob '*$name*'
order by
  registrationTime desc
limit 1
SQL
)))

# list all files in the bootstrap-archive accessed after the reference
# file
find $(sqlite3 -readonly /nix/var/nix/db/db.sqlite  <<SQL
select
  path
from
  ValidPaths
where
  path glob '*$name*'
order by
  registrationTime desc
limit 1
SQL
) -type f -anewer reference

```

dtrace could be nice, i haven't used it much tho seems like would need to find the different syscalls which open files (open, openat, etc) and then hope they are using absolute paths.

anyway, i agree this is not terribly important for this PR. the bootstrap size is already trimmed down from 1G+ to whatever, 200ish M and 40ish M compressed. seems like the fruit's been picked.